R/json_nest.R

```diff
@@ -24,11 +24,127 @@ json_nest <- function(.data, ..., .names_sep = NULL) {
+  tbl(dbplyr::remote_con(data), sql(query))
+}
+
+sql_json_nest <- function(con, cols, names_sep, packed_col, id_cols, data) {
+  UseMethod("sql_json_nest")
+}
+
+#' @export
+sql_json_nest.PqConnection <- function(con, cols, names_sep, packed_col, id_cols, data) {
+  inside_cols <- remove_prefix_and_sep(cols, prefix = packed_col, sep = names_sep)
+  inside_cols_idented <- dbplyr::ident(inside_cols)
+  exprs <- vctrs::vec_interleave(as.list(inside_cols_idented), syms(cols))
+  dbplyr::translate_sql(JSON_AGG(JSON_BUILD_OBJECT(!!!exprs)), con = con)
+}
+
+#' @export
+`sql_json_nest.Microsoft SQL Server` <- function(con, cols, names_sep, packed_col, id_cols, data) {
+  inside_cols <- remove_prefix_and_sep(cols, prefix = packed_col, sep = names_sep)
+  join_subquery <- glue_collapse(glue('"{id_cols}" = PLACEHOLDER."{id_cols}"'), " AND ")
+  filter_select_subquery <-
+    data %>%
+    filter(sql(join_subquery)) %>%
+    select(!!!set_names(syms(cols), inside_cols)) %>%
+    dbplyr::sql_render()
+  query <- glue("({filter_select_subquery} FOR JSON PATH)")
+  sql(query)
+}
+
+#' @export
+`json_nest.tbl_Microsoft SQL Server` <- function(.data, ..., .names_sep = NULL) {
+  # FIXME: we may not need json_nest.tbl_lazy if we implement json_nest methods for each DBMS
+  # FIXME: The old code is still there, just not called
+  # FIXME: `DBI::dbQuoteLiteral()` uses simple quotes for mssql, this seems wrong,
+  #   queries work if double quotes are used. should we implement our own fix ?
+  # FIXME: We need a table alias and we use `*tmp*`, can we leverage the mechanism
+  #   in `dbplyr` to increment the `q*` aliases ?
+
+  dots <- quos(...)
+  if ("" %in% names2(dots)) {
+    abort("All elements of `...` must be named.")
+  }
+  con <- dbplyr::remote_con(.data)
+
+  # define and protect column names, table name and alias
+  tbl_name <-
+    dbplyr::remote_name(.data) %>%
+    DBI::dbQuoteLiteral(con, .)
+  col_nms <- colnames(.data)
+  nest_cols <- purrr::map(dots, ~ tidyselect::vars_select(col_nms, !!.x))
+  id_cols <-
+    setdiff(col_nms, unlist(nest_cols)) %>%
+    DBI::dbQuoteLiteral(con, .)
+  col_nms <- DBI::dbQuoteLiteral(con, col_nms)
+  temp_alias <- DBI::dbQuoteLiteral(con, "*tmp*")
+
+  # build joining clause to use in final query
+  joins <- glue("{id_cols} = {temp_alias}.{id_cols}") %>%
+    glue_collapse(" AND ")
+
+  # compute subqueries for each nested column
+  nest_col_queries <- imap_chr(nest_cols, ~ {
+    cols <- toString(DBI::dbQuoteLiteral(con, .x))
+    alias <- DBI::dbQuoteLiteral(con, .y)
+    glue::glue("(SELECT {cols} FROM {tbl_name} WHERE ({joins}) FOR JSON PATH) AS {alias}")
+  })
+
+  # build final query
+  query <- glue::glue(
+    'SELECT {toString(id_cols)}, {toString(nest_col_queries)} ',
+    'FROM (SELECT {toString(col_nms)} FROM {tbl_name}) {temp_alias} GROUP BY {id_cols}'
+  )
+
+  # HACK! see FIXME at top
+  query <- gsub("'", '"', query)
+
+  tbl(con, sql(query))
```

Comment:
```suggestion
  tbl(con, sql(query), vars = ...)
```

to avoid an extra query to the database.